### ISSUE TITLE
After every overwrite & reload, it reloads DLL again after `debounce_duration`, even though it wasn't overwritten again

### ISSUE THREAD

**Issue by Boscop** [emoon/dynamic_reload#27]
I noticed some weird behavior:
On Windows 10, no matter which debounce duration I specify, whenever I overwrite the watched DLL (also tested with manual copying in Windows explorer), it reloads the DLL twice, once immediately and once after `debounce_duration`.
When I log the calls, I get this sequence:
```sh
16:49:47 [INFO] load_lib # Initial load_lib call
# I copy & overwrite the file in Windows Explorer, it reloads immediately
16:50:09 [INFO] reload_callback
16:50:09 [INFO] unload_lib
16:50:09 [INFO] reload_callback
16:50:09 [INFO] load_lib
16:50:09 [INFO] DLL reloaded
# Ten seconds later, it reloads again, even though the DLL wasn't overwritten again!
16:50:19 [INFO] reload_callback
16:50:19 [INFO] unload_lib
16:50:19 [INFO] reload_callback
16:50:19 [INFO] load_lib
16:50:19 [INFO] DLL reloaded
```
This is my code:
```rs
use dynamic_reload::{DynamicReload, Lib, PlatformName, Search, UpdateState};
use std::{path::Path, sync::Arc, time::Duration};

pub trait TypedLibApi: Sized {
	fn load(lib: &libloading::Library) -> Result<Self, libloading::Error>;
	fn before_unload(&self);
}

#[derive(SmartDefault)]
struct LibManager<T: TypedLibApi> {
	loaded_libs: Option<(Arc<Lib>, T)>,
	lib_was_loaded: bool,
}

impl<T: TypedLibApi> LibManager<T> {
	fn load_lib(&mut self, lib: &Arc<Lib>) {
		info!("load_lib");
		let api = T::load(&lib.lib).expect("TypedLibApi::load");
		self.loaded_libs = Some((lib.clone(), api));
		self.lib_was_loaded = true;
	}

	fn unload_lib(&mut self, lib: &Arc<Lib>) {
		info!("unload_lib");
		let (loaded_lib, api) = self.loaded_libs.take().unwrap();
		debug_assert!(loaded_lib == *lib);
		api.before_unload();
	}

	fn reload_callback(&mut self, state: UpdateState, lib: Option<&Arc<Lib>>) {
		info!("reload_callback");
		match state {
			UpdateState::Before => Self::unload_lib(self, lib.unwrap()),
			UpdateState::After => Self::load_lib(self, lib.unwrap()),
			UpdateState::ReloadFailed(e) => panic!("Failed to reload: {}", e),
		}
	}

	pub fn lib(&self) -> &T {
		&self.loaded_libs.as_ref().unwrap().1
	}
}

pub struct LibAutoReloader<T: TypedLibApi> {
	reload_handler: DynamicReload,
	lib_mgr: LibManager<T>,
}

impl<T: TypedLibApi> LibAutoReloader<T> {
	pub fn new(lib_path: impl AsRef<Path>, debounce_ms: u64) -> dynamic_reload::Result<Self> {
		let lib_path = lib_path.as_ref();
		let folder = lib_path.parent().unwrap().display().to_string();
		let file_name = lib_path.file_name().unwrap().to_str().unwrap();

		let shadow_dir = std::env::var_os("TEMP").unwrap().into_string().unwrap();
		let mut reload_handler = DynamicReload::new(
			Some(vec![&folder]),
			Some(&shadow_dir),
			Search::Default,
			Duration::from_millis(debounce_ms),
		);
		let mut lib_mgr = LibManager::default();
		match unsafe { reload_handler.add_library(file_name, PlatformName::No) } {
			Ok(lib) => lib_mgr.load_lib(&lib),
			Err(e) => {
				error!("Unable to load dynamic lib {}: {}", lib_path.display(), e);
				return Err(e);
			}
		}
		Ok(LibAutoReloader { reload_handler, lib_mgr })
	}

	pub fn reload_lib_if_changed(&mut self) -> bool {
		self.lib_mgr.lib_was_loaded = false;
		unsafe { self.reload_handler.update(&LibManager::reload_callback, &mut self.lib_mgr) };
		if self.lib_mgr.lib_was_loaded {
			info!("DLL reloaded");
		}
		self.lib_mgr.lib_was_loaded
	}

	pub fn lib(&self) -> &T {
		self.lib_mgr.lib()
	}
}
```
Then I instantiate it with `LibAutoReloader::new(&dll_path, 10_000)` and call `reload_lib_if_changed` regularly.

Any idea why this strange behavior happens? No matter how large I make the `debounce_duration`, it always reloads the DLL a second time exactly `debounce_duration` after it was actually overwritten!

I noticed this crate still uses notify-rs 4. Maybe this behavior would be fixed by upgrading to notify-rs 5.0.0 (which changed the debounce mechanism)? https://github.com/emoon/dynamic_reload/issues/28

**Comment by emoon** [comment]
Interesting. If you run the test(s) locally on your machine do you get the same behavior or is it in your program only?

**Comment by Boscop** [comment]
@emoon Thanks for the quick reply!

By tests do you mean the example?

I just tested it: I first added `println!("reload_plugin");` to the `reload_plugin` method to see when it reloads:
```rs
    fn reload_plugin(&mut self, lib: &Arc<Lib>) {
        Self::add_plugin(self, lib);
        println!("reload_plugin");
    }
```
Then I ran `cargo run --example example` in one cmd window, then changed the return value of `shared_fun` to 9999 and ran `cargo build` in another cmd window. 
This is the output:
```
Value 42
Value 42
Value 42
Value 42
Value 42
reload_plugin
Value 9999
Value 9999
Value 9999
Value 9999
reload_plugin
Value 9999
Value 9999
Value 9999
```
The second `reload_plugin` line was printed 2 seconds after the first (the debounce time in the example is `Duration::from_secs(2)`).
When I changed it to `Duration::from_secs(10)`, it reloaded again 10 seconds later. (So the second reload always happens after the debounce time.)

It would be interesting if you get the same behavior on your machine.

**Comment by emoon** [comment]
Sorry for the slow reply, I have been really busy. I will try to test this as soon as I can.

**Comment by emoon** [comment]
Hey, sorry for the very long delay. I tested this locally and I see the same behavior. I will try to investigate this as soon as I can

**Comment by Boscop** [comment]
Thanks! :)
It could be that upgrading to notify-rs 5.0 would fix this, I'm not sure but they changed the debouncing mechanism (and they moved it to a [separate crate](https://crates.io/crates/notify-debouncer-mini)).

**Comment by emoon** [comment]
0.10.0 has been released which includes uses `notify 5.1` and the separate Debouncer crate.
